lib/base.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * Docker client model
+ * @module lib/index
+ */
+'use strict'
+
+const Promise = require('bluebird')
+const defaults = require('101/defaults')
+const dogerode = require('dogerode')
+const assign = require('101/assign')
+const fs = require('fs')
+const join = require('path').join
+const url = require('url')
+
+const logger = require('./logger')
+
+const containerActions = [
+  'top',
+  'start',
+  'commit',
+  'inspect',
+  'stop',
+  'pause',
+  'unpause',
+  'restart',
+  'resize',
+  'attach',
+  'remove',
+  'copy',
+  'kill',
+  'exec',
+  'rename',
+  'log',
+  'stats',
+  'getArchive',
+  'infoArchive',
+  'putArchive',
+  'update'
+]
+
+// try/catch is a better pattern for this, since checking to see if it exists
+// and then reading files can lead to race conditions (unlikely, but still)
+const certs = {}
+
+try {
+  // DOCKER_CERT_PATH is docker's default thing it checks - may as well use it
+  const certPath = process.env.DOCKER_CERT_PATH || '/etc/ssl/docker'
+  certs.ca = fs.readFileSync(join(certPath, '/ca.pem'))
+  certs.cert = fs.readFileSync(join(certPath, '/cert.pem'))
+  certs.key = fs.readFileSync(join(certPath, '/key.pem'))
+} catch (e) {
+  throw e
+}
+
+/**
+ * Base Docker/Swarm client with batteries. Can be extended
+ * @param {Class} DriverClass docker driver class
+ * @param {Object} opts - docker options
+ * @class
+ * @extends loki:base
+ * @author Anton Podviaznikov
+ */
+class BaseDockerClient {
+
+  constructor (DriverClass, opts) {
+    if (!DriverClass) {
```

Comment:
no `joi`? fell like this would be better expresses as a `joi` schema. will make input more strict